packages/vitest/src/node/hoistMocks.ts

```diff
@@ -71,58 +72,114 @@ export function hoistMocks(code: string, id: string, parse: PluginContext['parse
-      code += `const ${namespace.local.name} = await import('${source}')\n`
-
-    // if we don't hijack ESM and process this file, then we definetly have mocks,
-    // so we need to transform imports into dynamic ones, so "vi.mock" can be executed before
-    const specifiers = transformImportSpecifiers(node)
-
-    if (specifiers) {
-      if (namespace)
-        code += `const ${specifiers} = ${namespace.local.name}\n`
-      else
-        code += `const ${specifiers} = await import('${source}')\n`
-    }
-    else if (!namespace) {
-      code += `await import('${source}')\n`
+    const importId = `__vi_import_${uid++}__`
+    const hasSpecifiers = node.specifiers.length > 0
+    const code = hasSpecifiers
+      ? `const ${importId} = await import('${source}')\n`
+      : `await import('${source}')\n`
+    return {
+      code,
+      id: importId,
     }
-    return code
   }
 
-  function hoistImport(node: Positioned<ImportDeclaration>) {
+  function defineImport(node: Positioned<ImportDeclaration>) {
     // always hoist vitest import to top of the file, so
     // "vi" helpers can access it
-    s.remove(node.start, node.end)
-
     if (node.source.value === 'vitest') {
       const code = `const ${transformImportSpecifiers(node)} = await import('vitest')\n`
       hoistedVitestImports += code
+      s.remove(node.start, node.end)
       return
     }
-    const code = transformImportDeclaration(node)
-    s.appendLeft(hoistIndex, code)
+
+    const declaration = transformImportDeclaration(node)
+    if (!declaration)
+      return null
+    s.appendLeft(hoistIndex, declaration.code)
+    return declaration.id
   }
 
   // 1. check all import statements and record id -> importName map
   for (const node of ast.body as Node[]) {
     // import foo from 'foo' --> foo -> __import_foo__.default
     // import { baz } from 'foo' --> baz -> __import_foo__.baz
     // import * as ok from 'foo' --> ok -> __import_foo__
-    if (node.type === 'ImportDeclaration')
-      hoistImport(node)
+    if (node.type === 'ImportDeclaration') {
+      const importId = defineImport(node)
+      if (!importId)
+        continue
+      s.remove(node.start, node.end)
+      for (const spec of node.specifiers) {
+        if (spec.type === 'ImportSpecifier') {
+          idToImportMap.set(
+            spec.local.name,
+            `${importId}.${spec.imported.name}`,
+          )
+        }
+        else if (spec.type === 'ImportDefaultSpecifier') {
+          idToImportMap.set(spec.local.name, `${importId}.default`)
+        }
+        else {
+          // namespace specifier
+          idToImportMap.set(spec.local.name, importId)
+        }
+      }
+    }
   }
 
+  const declaredConst = new Set<string>()
+
+  esmWalker(ast, {
+    onIdentifier(id, parent, parentStack) {
+      const grandparent = parentStack[1]
+      const binding = idToImportMap.get(id.name)
+      if (!binding)
+        return
+
+      if (isStaticProperty(parent) && parent.shorthand) {
```

Comment:
I feel like these if statements are duplicated here and in browser. Maybe insteadof `parentStack` we can pass down some kind of options object? `hasBindingShortcut: true, classDeclaration: false, classExpression: false` (don't forget to add comments with example why we are doing g the transform like in property shorthand)

Reply:
Sounds good